File: noise/noising.py
from PIL import Image
import imagehash, os
import numpy as np
import pandas as pd

SIZE=8
names = []

for c in ["f","m","o"]:
    li = []
    if(c=="m" or c=="f"):
        li=[1,2]
    else:
        li=[1,2,3,4]
    for i in li:
        ns=c+str(i)
        names.append(ns)
print(names)
# quit()

unif = {}
gaus = {}

for name in names:
    unif[name]=[]
    gaus[name]=[]


for name in names:
    for noise in range(20,241,20):
        basei = f'C:/Users/HP/Downloads/work/{name}/{name}.jpg'
        unifi = f'C:/Users/HP/Downloads/work/{name}/{name}u_{noise}.jpg'
        gausi = f'C:/Users/HP/Downloads/work/{name}/{name}g_{noise}.jpg'

        img = imagehash.phash(Image.open(basei),hash_size=SIZE)
        uni = imagehash.phash(Image.open(unifi),hash_size=SIZE)
        gas = imagehash.phash(Image.open(gausi),hash_size=SIZE)

        udst = img - uni
        gdst = img - gas

        unif[name].append(udst)
        gaus[name].append(gdst)

    print(f'{name} processed!')

print(unif)

udf = pd.DataFrame(unif)
gdf = pd.DataFrame(gaus)

udf.to_csv('C:/users/hp/downloads/py_u.csv',index=False)
gdf.to_csv('C:/users/hp/downloads/py_g.csv',index=False)

File: noise/noising.scala
import java.io.{FileInputStream, FileOutputStream, PrintWriter}
import java.nio.file.{Files, Paths}
import java.awt.image.BufferedImage
import javax.imageio.ImageIO
import scala.collection.mutable


object ImagePHashScalaExample extends App {
  val SIZE = 8
  val names = mutable.ArrayBuffer[String]()
  val order = List("f1", "f2", "m1", "m2", "o1", "o2", "o3", "o4")

  for (c <- List("f", "m", "o")) {
    val li = c match {
      case "m" | "f" => List(1, 2)
      case "o" => List(1, 2, 3, 4)
    }
    for (i <- li) {
      val ns = c + i.toString
      names += ns
    }
  }
  println(names)

  val unif = mutable.Map[String, List[Int]]()
  val gaus = mutable.Map[String, List[Int]]()

  for (name <- names) {
    unif(name) = List()
    gaus(name) = List()
  }

  val pHash = new ImagePHashScala(size = 64, smallerSize = 8)

  for (noise <- 20 to 240 by 20) {
    for (name <- names) {
      val basei = s"C:/Users/HP/Downloads/work/$name/${name}.jpg"
      val unifi = s"C:/Users/HP/Downloads/work/$name/${name}u_$noise.jpg"
      val gausi = s"C:/Users/HP/Downloads/work/$name/${name}g_$noise.jpg"

      val timg = new FileInputStream(basei)
      val tuni = new FileInputStream(unifi)
      val tgas = new FileInputStream(gausi)

      val img = pHash.getHash(timg)
      val uni = pHash.getHash(tuni)
      val gas = pHash.getHash(tgas)

      val udst = pHash.hammingDistance(img, uni)
      val gdst = pHash.hammingDistance(img, gas)

      unif(name) = unif(name) :+ udst
      gaus(name) = gaus(name) :+ gdst
    }
    println(s"Noise level $noise processed!")
  }

  saveToCSV("C:/users/hp/downloads/scl_u.csv", unif)
  saveToCSV("C:/users/hp/downloads/scl_g.csv", gaus)

  def saveToCSV(filePath: String, data: mutable.Map[String, List[Int]]): Unit = {
    val writer = new PrintWriter(new FileOutputStream(filePath))
    val headers = names.mkString(",") // Headers in the desired order
    writer.println(headers)

    // Write rows, each representing a noise level
    for (rowIdx <- data(names.head).indices) {
      val row = names.map(name => data(name)(rowIdx))
      writer.println(row.mkString(","))
    }
    writer.close()
  }
}